### PR TITLE
pkg: add a function to generate chain names

### PR DESCRIPTION
**Pull request by achanda**
Adds a function to generate chain names for use in iptables and
ports all drivers to use that function. Also adds tests for the
said function.
Fixes https://github.com/coreos/rkt/issues/2314

### CODE REVIEW

File: pkg/utils/utils.go [hunk 1]
```diff
@@ -0,0 +1,17 @@
+package utils
+
+import (
+        "crypto/sha512"
+        "fmt"
+)
+
+const maxChainName = 29
+
+func FormatChainName(name string, id string) string {
```
Comment (steveeJ) [inline pkg/utils/utils.go]: missing docstring

**Comment by steveeJ** [comment]
Thanks!

Please adjust your commit messages according to the [contribution guidelines about commit messages](https://github.com/appc/cni/blob/master/CONTRIBUTING.md#format-of-the-commit-message).

**Comment by achanda** [comment]
Thanks @steveeJ . I updated the commit.

**Comment by steveeJ** [comment]
I'm currently not up to thinking this through myself, so I'd rather ask this than forget about it.
Does this increase the chance of a chain name clash? If so, what happens in that case?

**Comment by achanda** [comment]
@steveeJ I do think this increases collision probability. `iptables` will error out in that case. Alternatives:
1. Disallow long network names
2. Format current system time as a 8 byte string to generate the name
Thoughts?

File: pkg/utils/utils_suite_test.go [hunk 2]
```diff
@@ -0,0 +1,13 @@
+package utils_test
+
```
Comment (alban) [inline pkg/utils/utils_suite_test.go]: The [TESTABLE](https://github.com/appc/cni/blob/master/test#L14) list must be updated.

File: pkg/utils/utils_test.go [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+
+import (
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+)
+
+var _ = Describe("Utils", func() {
+	It("should format a short name", func() {
+		chain := formatChainName("test", "1234")
```
Comment (alban) [inline pkg/utils/utils_test.go]: `formatChainName` with a uppercase "F".

File: plugins/main/bridge/bridge.go [hunk 4]
```diff
@@ -220,7 +221,7 @@ func cmdAdd(args *skel.CmdArgs) error {
 	}
 
 	if n.IPMasq {
-		chain := "CNI-" + n.Name
+		chain := utils.FormatChainName(n.Name, n.Name)
```
Comment (alban) [inline plugins/main/bridge/bridge.go]: Why not using `args.ContainerID` as in the plugin ptp?

File: pkg/utils/utils.go [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+)
+
+const maxChainName = 29
+
+// Generates a chain name to be used with iptables.
+// Ensures that the generated name is less than
+// maxChainName chars in length
+func FormatChainName(name string, id string) string {
+        h := sha512.Sum512([]byte(id))
+        chain := fmt.Sprintf("CNI-%s-%x", name, h[:8])
```
Comment (alban) [inline pkg/utils/utils.go]: What about using `name[:16]` to ensure that `h` will not be completely trimmed if `name` is very long?

File: pkg/utils/utils_test.go [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+		Expect(chain).To(Equal("CNI-test-d404559f602eab6f"))
+	})
+
+	It("should truncate a long name", func() {
+		chain := formatChainName("testtesttesttest", "1234")
+		Expect(chain).To(Equal("CNI-testtesttesttest-d404559f"))
+	})
+})
```
Comment (alban) [inline pkg/utils/utils_test.go]: you could also add another test for names longer than 29

**Comment by achanda** [comment]
@steveeJ the earlier implementation could cause collisions more frequently when network names are long. The new implementation does not change collision probability. Thoughts on this?

File: pkg/utils/utils_suite_test.go (hunk 2 above)
Comment (achanda) [inline pkg/utils/utils_suite_test.go]: Done now, thanks.

File: pkg/utils/utils.go [hunk 7]
```diff
@@ -0,0 +1,20 @@
...
+// 29 - len('CNI') - 2*len('-')
+const maxNameLen = 16
+
+// Generates a chain name to be used with iptables.
+// Ensures that the generated name is less than
+// 29 chars in length
+func FormatChainName(name string, id string) string {
+	h := sha512.Sum512([]byte(id))
+	if len(name) > maxNameLen {
+		return fmt.Sprintf("CNI-%s-%x", name[:len(name)-maxNameLen], h[:8])
```
Comment (steveeJ) [inline pkg/utils/utils.go]: @achanda 
This looks reasonable to me, please take a second look @alban

Comment (alban) [inline pkg/utils/utils.go]: Not sure to understand... shouldn't it be `name[:maxNameLen]`?

Also, to distinguish "verylongnameofthenetwork01" and "verylongnameofthenetwork02" (both truncated to "verylongnameofth"), could the hash `h` be feed with both `name` and `id`?

Comment (steveeJ) [inline pkg/utils/utils.go]: Looking at this closer I don't like the math, disregarding of the variable length, truncating a hash value will always increase the risk of collisions significantly. We need a hash function that produces an string output of 25 characters.

Comment (steveeJ) [inline pkg/utils/utils.go]: @achanda @alban PTAL at https://github.com/appc/cni/pull/168/files#diff-fb81b0a96eed5b010cfb081b3aef4431R15. It's a step closer but theoretically definitely not unique.

Comment (achanda) [inline pkg/utils/utils.go]: That's what I was trying to arrive at. Should we close this one then?

Comment (steveeJ) [inline pkg/utils/utils.go]: > Should we close this one then?

I think we can by now, the other PR is ready.

**Pull request closed without merging**